Tokio 0.2 blog post (#356)
tokio-rs/website: content/blog/2019-11-tokio-0-2.md

+++
date = "2019-11-26"
title = "Announcing Tokio 0.2 and a Roadmap to 1.0"
description = "November 26, 2019"
menu = "blog"
weight = 985
+++

We are very excited to announce Tokio 0.2. This is a ground up rework of Tokio
based on `async / await` and experience gained over the past three years.

Add the following to `Cargo.toml`:

```toml
tokio = { version = "0.2", features = ["full"] }
```

[API docs]: https://docs.rs/tokio/0.2

This is a major update, and as such there are changes to almost all parts of the
library. To highlight a few:

* Based on `async / await` for superior ergonomics.

> * A brand new, much faster, scheduler.

Reviewer: Maybe link to the scheduler blog post?

* Focus on making the Tokio dependency as lightweight as possible.

### `async / await`

If you have used Tokio to date, then you are familiar with the "old" way of
writing asynchronous Rust... it was not fun. Most ended up writing state
machines by hand. This was cumbersome and error prone.

As of Rust 1.39, the `async / await` is available on the stable channel and
Tokio 0.2 takes full advantage of it. Whenever possible, Tokio provides an
`async fn` based API, usually modeled after `std`.

For example, accepting sockets from a `TcpListener` is done with the async
`accept` function:

```rust,ignore

> let (mut socket, _) = listener.accept().await?;

Reviewer 2: ```suggestion
let (mut socket, _peer_addr) = listener.accept().await?;
```

Just to clarify what the value is to anyone unfamiliar.

```rust,ignore
let n = socket.read(&mut buf).await?;

let n = socket.write(&buf).await?;
```

The same applies to the rest of the Tokio API surface.

**Starting the Tokio runtime**

Setting the entry point of a Tokio application can now be done with a proc macro:

```rust,ignore
#[tokio::main]
async fn main() {
    println!("Hello from Tokio!");
}
```

This starts up the Tokio runtime and all necessary infrastructure to power the
application.

### A new scheduler

The new Tokio comes with a scheduler that was built from the ground up to take
advantage of the new asynchronous task system. It is based on the experience
gained from Tokio 0.1 as well as all the hard work put into **other** ecosystems
like Go, Erlang, Java, and others.

There are still improvements to make, but initial testing using [Hyper] shows a
30+% speed up in macro level benchmarks between the old scheduler and the new
scheduler.

You can read more about it [here][sched].

Since it landed on master, some Tokio users have been experimenting with the new
scheduler and have seen some very impressive real world improvements in their
applications. Hopefully they will blog about them soon!

> 

Reviewer 2: ```suggestion

[Hyper]: https://hyper.rs
```

number of transitive dependencies and added time spent compiling.

For example, a "hello world" Tokio 0.1 application on my laptop would pull in 43
crates and take 50 seconds to compile (not counting time spent downloading the
dependencies).

Tokio addresses this issue on two fronts. First, most Tokio components have been
collapsed into a single crate: `tokio` and transitive dependencies were pruned
aggressively. Second, Tokio components have been made **opt-in** using feature
flags instead of always on. Simply pulling in the `tokio` dependency will only
get you a few traits.

To get started with Tokio 0.2, you will need to request feature flags. A `full`
feature flag includes everything and is an easy way to get started:

```toml
tokio = { version = "0.2", features = ["full"] }
```

On my laptop, this reduces the total number of crates to 23. Compile time only
drops to 40 seconds.

Real benefits start happening when the user of Tokio starts only requesting the
components that are needed to run the application. To run the TCP echo server
example, the `io-util`, `rt-threaded`, and `tcp` feature flags are needed. Now,

> Tokio pulls in 13 crates and compiling takes **13 seconds**.

Reviewer: ```suggestion
Tokio pulls in 13 crates and compiles in **13 seconds**.
```

* [@taiki-e](https://github.com/taiki-e)

A pre-emptive thanks to all those working on Mio 0.7, which sadly didn't make
the cut for this release, but will happen soon!

* [@dtacalau](https://github.com/dtacalau)
* [@kleimkuhler](https://github.com/kleimkuhler)
* [@PerfectLaugh](https://github.com/PerfectLaugh)
* [@piscisaureus](https://github.com/piscisaureus)
* [@Thomasdezeeuw](https://github.com/Thomasdezeeuw)

And a big thanks to [Buoyant], the makers of [Linkerd] (the proxy is written in
Rust), who sponsored most of the work.

[Buoyant]: https://buoyant.io/
[Linkerd]: https://github.com/linkerd/linkerd2

## A Roadmap to 1.0

All that being said, we are shipping version 0.2. The question of "why not 1.0?"
has come up a few times. After all, Tokio 0.1 has been stable for three years.
The short answer: because it isn't time. There is nobody who would rather ship a
Tokio 1.0 than us. It also isn't something to rush.

After all, `async / await` only landed in the stable Rust channel weeks ago.
There has been no significant production validation yet, except maybe fuchsia
and that seems like a fairly specialized use case. This release of Tokio

> include significant new code and new strategies with feature flags. Also, there

Reviewer 2: ```suggestion
includes significant new code and new strategies with feature flags. Also, there
```

are still big open questions, such as the [proposed changes][pr-1744] to
`AsyncRead` and `AsyncWrite`.

Tokio 1.0 will be released as soon as the APIs are proven to handle real-world
production cases.

### Tokio 1.0 in Q3 2020 with LTS support

The Tokio 1.0 release will be **no later** than Q3 2020. It will also come with
"long-term support" guarantees:

* A minimum of 5 years of maintenance.
* A minimum of 3 years before a hypothetical 2.0 release.

When Tokio 1.0 is released in Q3 2020, on-going support, security fixes, and
critical bug fixes are guaranteed until **at least** Q3 2025. Tokio 2.0 will not

> be released until **at least** Q3 2023 (though, ideally there will never been a

Reviewer: ```suggestion
be released until **at least** Q3 2023 (though, ideally, there will never need to be a
```
